Revise and simplify NodePool's KubeVirt platform API

davidvossel opened the pull request:
## Overview

Previously, the KubeVirt platform's NodePool api consisted of the entire KubeVirt VM spec. We started with that because it gave us the freedom to quickly adapt our configurations as we experimented with various network and storage tunings. This does not provide a reasonable user experience though.

Now that we've gained more experience with the project, we feel comfortable pairing down the KubeVirt platform API to a small handful of tunings that are more approachable to end users.

## API Structure

The goal of this api revision is to provide a simplified api for configuring KubeVirt VMs within a Hypershift NodePool that only exposes the bare minimal tuning users need to be aware of when using the KubeVirt platform. We've also attempted to structure this API in a way that allows us to expand easily in the future, which is why there is some nested structure patterns that might not seem optimal at first glance.

For now, the API only has two top level structs `compute` and `rootVolume`. The compute structure lets users define things like the number of cores and amount of memory to give the guest. The `rootVolume` structure is where users define where the rhcos images comes from and the type of storage to use (only persistent is supported right now, but this can be expanded to ephemeral storage in the future)

In the future we might have a new top level struct dedicated to network configurations once more advanced tunings are supported there. Also, the compute struct might get the addtion of a VM `flavor` to define the mem/cpu usage for the guest once Openshift Virtualization officially supports flavors.

More info related to the api structure can be found in [this document](https://docs.google.com/document/d/17vMxzPZuq-XvHkkTp6E-CPdopMdNqOk48_0e61G_Pmg/edit#heading=h.y4i6eo4m47u3). 

## Notable Changes

With this api change, we've transitioned the kubevirt platform to only use persistent storage (VM's backed by PVCs). As a result, we've introduced two new hypershift CLI options, `--root-volume-size` and `--root-volume-storage-class`. Neither values have are required, but offer users the ability to customize the root volume size and the storage class that will provision the volumes.

## Example NodePool with KubeVirt Platform

```
spec:
    clusterName: example-6thkl
    management:
      autoRepair: false
      replace:
        rollingUpdate:
          maxSurge: 1
          maxUnavailable: 0
        strategy: RollingUpdate
      upgradeType: Replace
    nodeCount: 5
    platform:
      kubevirt:
        compute:
          cores: 2
          memory: 4Gi
        rootVolume:
          diskImage:
            containerDiskImage: quay.io/dvossel/rhcos:4.11-pre-release
          persistent:
            size: 16Gi
      type: KubeVirt
    release:
      image: quay.io/openshift-release-dev/ocp-release:4.11.0-0.nightly-multi-2022-04-19-151226-x86_64
```

davidvossel commented:
/cc @nirarg

davidvossel commented:
/retest

nirarg commented:
/lgtm

nirarg commented:
/retest

nirarg commented:
/lgtm

enxebre commented on hypershift-operator/controllers/nodepool/kubevirt.go:
```diff
@@ -1,39 +1,168 @@
 )
 
+const (
+	defaultMemory       = "4Gi"
```
why is all this defaulted in the backend instead of in the API?

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -413,11 +413,67 @@ type PowerVSNodePoolPlatform struct {
+type KubevirtCompute struct {
+	// Memory represents how much guest memory the VM should have
+	//
+	// +optional
```
I assumed some of these optional values are necessary for the controller to work. Why arent' we defaulting values here and documenting it? so as a consumer when I look at the API docs I'll see something like // Memory represents how many cores the guest VM should have. If unset the default value is x

davidvossel replied:
i added kubebuilder defaults and better documented the defaults

enxebre commented:
nit: could we include the PR desc as the commit desc as well?

enxebre commented on hypershift-operator/controllers/nodepool/kubevirt.go:
```diff
@@ -1,39 +1,168 @@
+const (
+	defaultMemory       = "4Gi"
+	defaultCores        = uint32(2)
+	defaultVolumeSizeGi = 16
```
have you considered using https://pkg.go.dev/k8s.io/apimachinery/pkg/api/resource to manage this input?

davidvossel replied:
The api is using a `resource` object. I just used an int here because it couldn't figure out how to make this default a const if it was a `resource` object. 

For example, this compiles.

```
var (
        defaultVolumeSize = apiresource.MustParse("16Gi")
)
```

this does not compile.

const (
        defaultVolumeSize = apiresource.MustParse("16Gi")
)

davidvossel commented:
@enxebre thanks for the review. I updated the PR to more clearly communicate the api defaults. I also made the API fill in most of the defaults using the kubebuilder markers.

I left the backend defaults in the code though. Here's my reasoning.
- Some of the defaults are difficult to express on the API regarding storage. In the future i plan on expanding the storage types to `ephemeral` and `persistent`. Someone would choose persistent or ephemeral storage by defining the struct they want to use. So the type of storage used would be defined as either the storage value that isn't `nil`, or use `persistent` storage (default) if both ephemeral and persistent are nil.
- I feel uneasy about the crd api defaulting. It's unclear to me if someone can force a value to `nil` in a way that won't result in a default being set on the object itself. Unless the backend knows how to handle that gracefully, we'd have to make the reconcile loop treat that as an error, which i'd rather avoid.

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +419,74 @@ type PowerVSNodePoolPlatform struct {
+	// +kubebuilder:default="4Gi"
+	Memory *resource.Quantity `json:"memory,omitempty"`
+
+	// Memory represents how many cores the guest VM should have
```
typo Memory -> Cores

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +419,74 @@ type PowerVSNodePoolPlatform struct {
+// KubevirtRootVolume represents the volume that the rhcos disk will be stored and run from.
+type KubevirtRootVolume struct {
+	// Image represents what rhcos image to use for the node pool
+	Image *KubevirtDiskImage `json:"diskImage,omitempty"`
```
as a omitempty pointer this is optional, though it'd be good to add the +optional for consistency/readibility.

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +419,74 @@ type PowerVSNodePoolPlatform struct {
+// KubevirtDiskImage contains values representing where the rhcos image is located
+type KubevirtDiskImage struct {
+	// ContainerDiskImage is a string representing the container image that holds the root disk
+	ContainerDiskImage *string `json:"containerDiskImage,omitempty"`
```
same thought than https://github.com/openshift/hypershift/pull/1314/files#r864557580

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +419,74 @@ type PowerVSNodePoolPlatform struct {
-	// NodeTemplate Spec contains the VirtualMachineInstance specification.
-	NodeTemplate *capikubevirt.VirtualMachineTemplateSpec `json:"nodeTemplate,omitempty"`
+	// RootVolume represents values associated with the VM volume that will host rhcos
+	RootVolume *KubevirtRootVolume `json:"rootVolume,omitempty"`
```
same thought than https://github.com/openshift/hypershift/pull/1314/files#r864557580

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +419,74 @@ type PowerVSNodePoolPlatform struct {
+	// Compute contains values representing the virtual hardware requested for the VM
+	//
+	// +optional
+	Compute KubevirtCompute `json:"compute,omitempty"`
```
why is this not a pointer but RootVolume is? I think this field should actually be required, how can otherwise the backend work without this?

enxebre commented:
>Someone would choose persistent or ephemeral storage by defining the struct they want to use. So the type of storage used would be defined as either the storage value that isn't nil, or use persistent storage (default) if both ephemeral and persistent are nil.

For this you'll want to use a union discriminator to signal the type and  expectations
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#automatic-resource-allocation-and-deallocation
https://github.com/kubernetes/kubernetes/issues/35345

Then the nodePool backend should validate the input is valid as defined above and fail in a condition otherwise.

enxebre commented on hypershift-operator/controllers/nodepool/kubevirt.go:
```diff
@@ -1,39 +1,167 @@
+	volumeSize := apiresource.MustParse(fmt.Sprintf("%vGi", defaultVolumeSizeGi))
+
+	if kvPlatform.Compute.Memory != nil {
+		memory = *kvPlatform.Compute.Memory
```
this should never be nil since we are defaulting in the API. However for robustness I'd expect this to be in a validate function which sets a failed condition if the values are empty.

davidvossel commented:
@enxebre thanks for the feedback. I think I've addressed everything in the latest round. Below are the big items i changed

- Compute and RootVolume are now required and both are pointers with defaults
- I transitioned to using a `unionDiscriminator` for the kubevirt volume types now... even though we only support `persistent` at the moment.
- I cleaned up usage of `optional` and `omitempty`... pretty much everything is required right now other than StorageClass in the persistent volume struct.
- I introduced some basic validation similar to what we have for aws and other platforms that ensures the union types and structs line up like we expect. example, type=kubevirt and kubevirtplatform is != nil,  type=Persistent and persistent!=nil. A validation error condition is set explaining why the nodepool is invalid if this occurs.
- no more backend defaulting. All defaulting is handled at the api level.

enxebre commented on hypershift-operator/controllers/nodepool/nodepool_controller.go:
```diff
@@ -423,6 +423,21 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
 
+	// Validate KubeVirt platform specific format
+	if nodePool.Spec.Platform.Type == hyperv1.KubevirtPlatform {
+		err := kubevirtPlatformValidation(nodePool)
```
nit: this if could be inlined.

davidvossel replied:
yep, np

enxebre commented on hypershift-operator/controllers/nodepool/nodepool_controller.go:
```diff
@@ -423,6 +423,21 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
+			return ctrl.Result{}, fmt.Errorf("validation of NodePool KubeVirt platform failed: %w", err)
+		}
+	}
+
```
We need to clean the condition here in case you go from failed to succeed, e.g removeStatusCondition(&nodePool.Status.Conditions, hyperv1.NodePoolValidKubevirtConfigConditionType)

davidvossel replied:
ah, good catch, fixed!

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -13,6 +13,7 @@ const (
 	NodePoolValidAMIConditionType                = "ValidAMI"
 	NodePoolValidPowerVSImageConditionType       = "ValidPowerVSImage"
 	NodePoolValidMachineConfigConditionType      = "ValidMachineConfig"
+	NodePoolValidKubevirtConfigConditionType     = "ValidKubevirtConfig"
```
To be consistent for consumers we should probably have a generic condition e.g. ValidProviderConfiguration where InvalidKubevirtConfig is the reason but will figure in follow up.

davidvossel replied:
Makes sense. I'd like to see a nodepool validation webhook similar to what we have for hostedcluster as well. I don't mind following up with this after this PR goes in. it probably does belong in another PR though since it will impact multiple platforms

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +420,81 @@ type PowerVSNodePoolPlatform struct {
+	RootVolume *KubevirtRootVolume `json:"rootVolume"`
+
+	// Compute contains values representing the virtual hardware requested for the VM
+	// +kubebuilder:default={memory: "4Gi", cores: 2}
```
Now this field is conceptually required, which makes sense. However if you want to use kubebuilder defaulting you must signal the field as `// +optional` as well (as we do for any other defaulted field).

The reason is that structural schema CRDs don't publish defaults so client-side can't know what it would be defaulted on the server. This results in a surprising UX for fields that are both required and defaulted failing client side validation if they are not set explicitly.

davidvossel replied:
interesting. I added +optional where ever defaulting is occurring now

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +420,81 @@ type PowerVSNodePoolPlatform struct {
+	// Memory represents how much guest memory the VM should have
+	//
+	// +kubebuilder:default="4Gi"
+	Memory *resource.Quantity `json:"memory"`
```
https://github.com/openshift/hypershift/pull/1314/files#r865699495

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +420,81 @@ type PowerVSNodePoolPlatform struct {
+	// Size is the size of the persistent storage volume
+	//
+	// +kubebuilder:default="16Gi"
+	Size *resource.Quantity `json:"size"`
```
https://github.com/openshift/hypershift/pull/1314/files#r865699495

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +420,81 @@ type PowerVSNodePoolPlatform struct {
+// KubevirtDiskImage contains values representing where the rhcos image is located
+type KubevirtDiskImage struct {
+	// ContainerDiskImage is a string representing the container image that holds the root disk
+	ContainerDiskImage *string `json:"containerDiskImage"`
```
unless you need to differentiate between nil and "" I don't think this should be a pointer.

davidvossel replied:
I'd rather keep this as a pointer. I'm considering making this a union in the future where someone could specify a disk from an HTTP endpoint, or in a PVC, which would all get treated differently

davidvossel commented:
@enxebre  update, ptal, thanks!

enxebre commented on api/v1alpha1/nodepool_types.go:
```diff
@@ -419,11 +420,86 @@ type PowerVSNodePoolPlatform struct {
+	//
+	// +unionDiscriminator
+	// +kubebuilder:default=Persistent
+	Type KubevirtVolumeType `json:"type"`
```
this also applies here https://github.com/openshift/hypershift/pull/1314/files#r865699495

enxebre commented on hypershift-operator/controllers/nodepool/nodepool_controller.go:
```diff
@@ -423,6 +423,22 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
+			})
+			return ctrl.Result{}, fmt.Errorf("validation of NodePool KubeVirt platform failed: %w", err)
+		} else {
+			removeStatusCondition(&nodePool.Status.Conditions, hyperv1.NodePoolValidKubevirtConfigConditionType)
```
nit: don't really need the else

enxebre commented:
thanks @davidvossel lgtm other than https://github.com/openshift/hypershift/pull/1314#discussion_r865960814

enxebre commented:
thanks @davidvossel!
/approve
/lgtm

davidvossel commented:
/retest-required

alvaroaleman commented:
/hold cancel
